src/transport/raw/TCP.cpp

```diff
@@ -216,56 +257,55 @@ CHIP_ERROR TCPBase::SendMessage(const Transport::PeerAddress & address, System::
     uint8_t * output = msgBuf->Start();
     LittleEndian::Write32(output, static_cast<uint32_t>(msgBuf->DataLength() - kPacketSizeBytes));
 
-    // Reuse existing connection if one exists, otherwise a new one
-    // will be established
-    ActiveTCPConnectionState * connection = FindActiveConnection(address);
-
-    if (connection != nullptr)
+    // Must find a previously-established connection with an owning reference
+    auto connection = FindInUseConnection(address);
+    VerifyOrReturnError(!connection.IsNull(), CHIP_ERROR_INCORRECT_STATE);
+    if (connection->IsConnected())
     {
         return connection->mEndPoint->Send(std::move(msgBuf));
     }
 
-    return SendAfterConnect(address, std::move(msgBuf));
+    return SendAfterConnect(connection, address, std::move(msgBuf));
 }
 
 CHIP_ERROR TCPBase::StartConnect(const PeerAddress & addr, Transport::AppTCPConnectionCallbackCtxt * appState,
-                                 Transport::ActiveTCPConnectionState ** outPeerConnState)
+                                 ActiveTCPConnectionHolder & outPeerConnState)
 {
 #if INET_CONFIG_ENABLE_TCP_ENDPOINT
-    ActiveTCPConnectionState * activeConnection = nullptr;
-    Inet::TCPEndPoint * endPoint                = nullptr;
-    *outPeerConnState                           = nullptr;
+    Inet::TCPEndPoint * endPoint = nullptr;
+    outPeerConnState.Release();
     ReturnErrorOnFailure(mListenSocket->GetEndPointManager().NewEndPoint(&endPoint));
 
-    auto EndPointDeletor = [](Inet::TCPEndPoint * e) { e->Free(); };
-    std::unique_ptr<Inet::TCPEndPoint, decltype(EndPointDeletor)> endPointHolder(endPoint, EndPointDeletor);
+    std::unique_ptr<Inet::TCPEndPoint, EndPointDeletor> endPointHolder(endPoint, EndPointDeletor());
 
-    endPoint->mAppState         = reinterpret_cast<void *>(this);
-    endPoint->OnConnectComplete = HandleTCPEndPointConnectComplete;
-    endPoint->SetConnectTimeout(mConnectTimeout);
+    InitEndpoint(endPoint);
 
-    activeConnection = AllocateConnection();
-    VerifyOrReturnError(activeConnection != nullptr, CHIP_ERROR_NO_MEMORY);
-    activeConnection->Init(endPoint, addr);
+    ActiveTCPConnectionHolder activeConnection = AllocateConnection(endPoint, addr);
```

Comment:
AllocateConnection actually returns a pointer to ActiveTCPConnectionState. For clarity, should activeConnection be  typed as such?